### PR TITLE
Add files via upload (docker volume article)

### PR DESCRIPTION
**Pull request by FireFrozen94**
this article plain to explain how to use bind mounts in docker, in the particular example of building a log system

### CODE REVIEW

**Comment by ctmbl** [comment]
hey @FireFrozen94 if you see this I've updated a bit the article I tjink it is ready for publishing 👍

**Comment by FireFrozen94** [comment]
Yay let's gooo

Le sam. 13 avr. 2024, 06:07, ClementMabileau ***@***.***> a
écrit :

File: src/content/posts/docker_volume_introduction.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+> We'll then **not distinguish one from the other**, some of what we say might better applied to bind mounts but **we'll only speak of volumes**.  
+> For more information see [Manage data in Docker](https://docs.docker.com/storage/).
+
+## What is a Docker's volume?
+
+A docker's volume is the action of **mounting an existing file or directory** of the host (the computer running the container) **into a container**. To rephrase, it enables a container to **share a memory space with the host**. So both the container and the user can access and modify it in real time and because it's not "contained" in the container but only mounted, it'll **persist after the container's death**.
+
+## Why do we need volumes?
+
+Docker containerization is very useful on lots of points for development but it's very annoying if your application **need to produce persistent data**, like logs. Indeed if you **kill** your container (to update it for example) or if it **crashes**, you **lose everything that where inside** and understanding why your application crashed without its logs can be very complicated or even impossible.
```
Comment (ZynoXelek) [inline src/content/posts/docker_volume_introduction.md:26]:
```suggestion
Docker containerization is very useful on lots of points for development but it's very annoying if your application **need to produce persistent data**, like logs. Indeed if you **kill** your container (to update it for example) or if it **crashes**, you **lose everything that was inside** and understanding why your application crashed without its logs can be very complicated or even impossible.
```

File: src/content/posts/docker_volume_introduction.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+    docker run --rm --interactive --tty \
+	--detach \
+	--volume $(realpath -P ${LOG_FOLDER}):/opt/${NAME}/logs \
+	--env-file .env.prod \
+	--name ${NAME} \
+	${NAME}:latest
+```
+In this example `LOG_FOLDER` is an environment variable defined in the `.env.prod` file.  
+The code will written transparently to `/opt/${NAME}/logs/<log file>` but because this folder is shared with the host, logs will **be available in the host** and survive potential application crashes or updates.
```
Comment (ZynoXelek) [inline src/content/posts/docker_volume_introduction.md:65]:
```suggestion
The code will write transparently to `/opt/${NAME}/logs/<log file>` but because this folder is shared with the host, logs will **be available in the host** and survive potential application crashes or updates.
```

Comment (ZynoXelek) [reply src/content/posts/docker_volume_introduction.md:65]: Or 'will be written' maybe ?

**Review by ZynoXelek — CHANGES_REQUESTED** [review]
Once this will be corrected, it seems good to be published to me!

File: src/content/posts/docker_volume_introduction.md (hunk 2 above)
Comment (ctmbl) [reply src/content/posts/docker_volume_introduction.md:65]: nop 'write' is expected indeed!

Comment (ctmbl) [reply src/content/posts/docker_volume_introduction.md]: [9ff9c64](https://github.com/iScsc/blog.iscsc.fr/pull/28/commits/9ff9c6474f45d0158cfe76218d6d932b68a02cd9)

File: src/content/posts/docker_volume_introduction.md (hunk 1 above)
Comment (ctmbl) [reply src/content/posts/docker_volume_introduction.md]: [9ff9c64](https://github.com/iScsc/blog.iscsc.fr/pull/28/commits/9ff9c6474f45d0158cfe76218d6d932b68a02cd9)

File: src/content/posts/docker_volume_introduction.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+> **Disclaimer**:  
+> Note that in Docker, **two ways exist to persist/share data** with the host: **volumes** and **bind mounts**.  
+> While some key differences exist between both, they are not relevant in most of our basic use cases.  
+> We'll then **not distinguish one from the other**, some of what we say might better applied to bind mounts but **we'll only speak of volumes**.  
+> For more information see [Manage data in Docker](https://docs.docker.com/storage/).
+
+## What is a Docker's volume?
+
+A docker's volume is the action of **mounting an existing file or directory** of the host (the computer running the container) **into a container**. To rephrase, it enables a container to **share a memory space with the host**. So both the container and the user can access and modify it in real time and because it's not "contained" in the container but only mounted, it'll **persist after the container's death**.
```
Comment (atxr) [inline src/content/posts/docker_volume_introduction.md:22]: > So both the container and the user can access and modify it in real time

Mmh that is right but they should not
According to [the link you put just above](https://docs.docker.com/storage/#choose-the-right-type-of-mount):

> - Volumes are stored in a part of the host filesystem which is managed by Docker (/var/lib/docker/volumes/ on Linux). Non-Docker processes should not modify this part of the filesystem. Volumes are the best way to persist data in Docker.
> - Bind mounts may be stored anywhere on the host system. They may even be important system files or directories. Non-Docker processes on the Docker host or a Docker container can modify them at any time.

You explained pretty well in the two sentences before so I would just remove this and rephrase like this:
> Because files inside the volumes are stored on the host, they will persist after the container's death.

File: src/content/posts/docker_volume_introduction.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+So we need to find a way to **keep certain files after the container's death** and that can be **achieved with volumes**!.
+
+## How to create a volume?
+
+To create a volume, you need to add a specific argument to the `docker run` command.
+
+> You can also create volumes outside of containers with [`docker volume`](https://docs.docker.com/storage/volumes/#create-and-manage-volumes)
+
+There is two different arguments which are really similar in their use. Those are `--mount` and `--volume`, both can be used to create volumes (or bind mounts, see Disclaimer above).
```
Comment (atxr) [inline src/content/posts/docker_volume_introduction.md:36]: I think you should wipe the `--mount` part
As you said in the disclaimer, `--volume` is more appropriate for your usage, so it's a bit confusing that you still talk about it here.

File: src/content/posts/docker_volume_introduction.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+The code will write transparently to `/opt/${NAME}/logs/<log file>` but because this folder is shared with the host, logs will **be available in the host** and survive potential application crashes or updates.
+
+> **WARNING : Be careful that the bot/application running in the container must have the right permission to modify the file/directory**
+
+## Other important use cases of volumes
+
+* The same volume can be shared by multiple containers.  
+ It can be useful if you want to share a database with applications running on different containers or to centralized logs of multiple application in the same log file/folder.
+* In development you can use volumes to share code with the container (where it runs) while editing it in your IDE in the host.
+* Volumes exist and can be created/managed outside of a container (contrary to bind mounts) with `docker volume`
```
Comment (atxr) [inline src/content/posts/docker_volume_introduction.md:74]: > Contrary to bind mounts

I'm a bit confused about this.
A bind mount is just a directory in the host file system that you mount in your container. So technically, it is always available outside of a container.

File: src/content/posts/docker_volume_introduction.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+In this example `LOG_FOLDER` is an environment variable defined in the `.env.prod` file.  
+The code will write transparently to `/opt/${NAME}/logs/<log file>` but because this folder is shared with the host, logs will **be available in the host** and survive potential application crashes or updates.
+
+> **WARNING : Be careful that the bot/application running in the container must have the right permission to modify the file/directory**
+
+## Other important use cases of volumes
+
+* The same volume can be shared by multiple containers.  
+ It can be useful if you want to share a database with applications running on different containers or to centralized logs of multiple application in the same log file/folder.
+* In development you can use volumes to share code with the container (where it runs) while editing it in your IDE in the host.
```
Comment (atxr) [inline src/content/posts/docker_volume_introduction.md:73]: In this case, bind mounts are a better option
See https://docs.docker.com/storage/#good-use-cases-for-bind-mounts

**Review by atxr — COMMENTED** [review]
Interesting topic!
Few comments but the overall quality of your article is really good!

**Comment by ctmbl** [comment]
@atxr tks for your comments, when I read them I got the feeling that you were putting the finger where it huts 😆 

But that's also for the best, to be honest I knew I was missing something about the distinction between volumes/bind mounts, you just gave me the opportunity to enlighten it and learn a bit more!

So let's dive into it (with further research and reading from me)!

> **DISCLAIMER**:  
> I said some (partially) wrong things or confusions before, and will for sure say some again  
> I'm just stating what I think is true at the moment given what I've read, understood and learnt in the hope it will help someone walk the same path, just faster 🙂 
> What I expose here you might have already understood but it wasn't clear to me so I'll write it anyway 🙂 

## TLDR
What we're talking about in this blog aren't volumes at all (even if we use `--volume`) they are bind mounts, we're just naming it volumes because we're confusing both.

**Volumes are not only managed by `-v` and bind mounts not only managed by `--mount`.  
Flags names and usage are just very very confusing**

Examples with `-v` but __**similar syntax is possible with `--mount`**__:
- bind mount: `docker run -v <path/to/host>:<path/to/cont>`
- named volume: `docker run -v <pre-created or not volume name>:<path/to/cont>`
- anonymous volum: `docker run -v <path/to/cont>` 

You'll see that even in [docker's doc the naming is confusing](https://docs.docker.com/storage/volumes/#differences-between--v-and---mount-behavior)
> [Differences between -v and --mount behavior](https://docs.docker.com/storage/volumes/#differences-between--v-and---mount-behavior)
> As opposed to bind mounts, all options for volumes are available for both --mount and -v flags.
> **Volumes used with services, only support --mount.**

**WTF**

## Detailed explanation

Actually there are **3** ways of sharing data in docker:
- bind mounts
- **named** volumes
- **anonymous** volumes

### Named volumes

They are a space which exists outside of the container, **managed by docker** (exist in `/var/lib/docker/volumes` but [you should never write to it form host](https://forums.docker.com/t/write-from-host-to-volume/47274)), and survives containers death.  
They have a name.

Named volumes are created either with the `docker volume` subcommand or with `docker run -v <some new volume name>:</path/in/cont>` **or similar with `--mount syntax`**

[about `docker volume`](https://docs.docker.com/storage/volumes/#create-and-manage-volumes)

[about `-v` in **volumes** doc](https://docs.docker.com/storage/volumes/#choose-the--v-or---mount-flag):
> -v or --volume: Consists of three fields, separated by colon characters (:). The fields must be in the correct order, and the meaning of each field isn't immediately obvious.
>   - **In the case of named volumes, the first field is the name of the volume**, and is unique on a given host machine. For anonymous volumes, the first field is omitted.

also [here](https://dev.to/doziestar/a-comprehensive-guide-to-docker-volumes-4d9h) at `Named Volumes` section

### Anonymous volumes

They are volumes too, only we don't give them a name.  
Their only interest (**to my understanding**) is to share space **across** containers, because they have no name they cannot be easily reused.

They are created with `docker run -v <path/in/cont>` and then referred to (because they have no name) and shared with another container with `docker run --volumes-from <first container id>` or similar with `--mount` syntax

[about their creation with `-v`](https://docs.docker.com/storage/volumes/#choose-the--v-or---mount-flag):
> -v or --volume: Consists of three fields, separated by colon characters (:). The fields must be in the correct order, and the meaning of each field isn't immediately obvious.
>   - In the case of named volumes, the first field is the name of the volume, and is unique on a given host machine. **For anonymous volumes, the first field is omitted.**

[example of `--volumes-from`](https://www.baeldung.com/ops/docker-volumes#3-using---volumes-from-to-share-volumes)

### Bind mounts

Bind mounts are actually what we describe in the article, they are a file or folder existing on host (at another path than `/var/lib/docker/*`) that is shared with the container, they can also be shared **across** containers because they can be mounted several time (but I don't know if this is a good practice honestly).

They are created with `docker run -v <path/to/host>:<path/to/cont>` or similar with `--mount syntax`.

[about `-v` in **bind mounts** doc](https://docs.docker.com/storage/bind-mounts/#choose-the--v-or---mount-flag):
> -v or --volume: Consists of three fields, separated by colon characters (:). The fields must be in the correct order, and the meaning of each field is not immediately obvious.
> -   **In the case of bind mounts,** the first field is the path to the file or directory on the host machine.


## Ressources

docker doc:
- [Choose the -v or --mount for volumes](https://docs.docker.com/storage/volumes/#choose-the--v-or---mount-flag)
- [Choose the -v or --mount for bind mounts](https://docs.docker.com/storage/bind-mounts/#differences-between--v-and---mount-behavior)
- [Differences between -v and --mount behavior](https://docs.docker.com/storage/volumes/#differences-between--v-and---mount-behavior)
- [Create and manage volumes](https://docs.docker.com/storage/volumes/#create-and-manage-volumes)

Other sources:
- https://dev.to/doziestar/a-comprehensive-guide-to-docker-volumes-4d9h
- https://www.baeldung.com/ops/docker-volumes
- [Never mess with `/var/lib/docker` docker forum](https://forums.docker.com/t/write-from-host-to-volume/47274)

File: src/content/posts/docker_volume_introduction.md (hunk 5 above)
Comment (ctmbl) [reply src/content/posts/docker_volume_introduction.md:74]: sorry it wasn't clear, I was trying to say that docker volumes and containers have completely separated existence and lifecyle

[docker doc](https://docs.docker.com/storage/volumes/#create-and-manage-volumes)
> Unlike a bind mount, you can create and manage volumes outside the scope of any container.

But my sentence isn't clear I'll change it!

File: src/content/posts/docker_volume_introduction.md (hunk 4 above)
Comment (ctmbl) [reply src/content/posts/docker_volume_introduction.md:36]: what I didn't understand before was that using `-v` doesn't necessarily creates volumes and using `--mount` doesn't necessarily creates bind mounts
see the TLDR of my answer below

File: src/content/posts/docker_volume_introduction.md (hunk 3 above)
Comment (ctmbl) [reply src/content/posts/docker_volume_introduction.md:22]: When I say `both the container and the user can access and modify it in real time` I mean the **corresponding file in the host** because we're presenting `-v <path/to/host>:<path/to/cont>` syntax (which isn't a volume but a bind mount strictly speaking), [**of course never touch anything in `/var/lib/docker`**](https://forums.docker.com/t/write-from-host-to-volume/47274)

**Comment by ctmbl** [comment]
@atxr What I've written above is almost a blog post in itself ahah

Sorry for the extensive writing, I just wanted to avoid any possible misunderstanding which would have necessarily lead to pointless debates. 
If you don't feel like digging in the details just read the TLDR section it should be enough to explain what was missleading us in my previous version of the article.

In the end it was @FireFrozen94 who was right to speak about bind mounts in his article, they are nothing alike volume and I was wrong to confuse them.  
**We can confuse `--volume` and `--mount` but we can't confuse the concepts of volumes and bind mounts**

I'll soon propose a version that should satisfy everyone

**Comment by atxr** [comment]
Don't worry I didn't dig dip enough to get all the details you just highlighted in your TLDR
I just spotted there was some misunderstanding about mounts and volumes that were "thrown under the carpet" :broom: 
I'm glad it helped!
Looking forward the fix! :waffle:

**Comment by ctmbl** [comment]
@atxr @ZynoXelek I finally found some time to propose an up-to-date and confusion-less (I hope) version of the article 

Sorry @FireFrozen94 for the previous modifications which were partially wrong, my bad 😓

File: src/content/posts/docker_volume_introduction.md [hunk 7]
```diff
@@ -1,84 +1,104 @@
...
+
+This article addresses the concept of bind mounts.
+
+---
+
+In Docker, two ways exist to persist/share data: **volumes** (named and anonymous) and **bind mounts**.  
+
+What we're talking about in this blog aren't volumes at all, they are bind mounts.  
+We'll name them *bind mounts* **but be aware that** often on Internet you'll encounter the word *volume* while the author is actually speaking of a bind mount.  
+This is certainly due to the fact that **options `-volume` and `--mount` can be used almost interchangeably**.
```
Comment (atxr) [inline src/content/posts/docker_volume_introduction.md:33]: --volume

**Review by atxr — COMMENTED** [review]
Wonderful :unicorn: 
Much clearer :bulb:

File: src/content/posts/docker_volume_introduction.md (hunk 7 above)
Comment (ctmbl) [reply src/content/posts/docker_volume_introduction.md:33]: indeed! tks!

Comment (ctmbl) [reply src/content/posts/docker_volume_introduction.md]: fix [5828f00](https://github.com/iScsc/blog.iscsc.fr/pull/28/commits/5828f003b8384ffe5f1fcb99a9304c5109723c03)

**Review by ZynoXelek — APPROVED** [review]
I think it is ready to be published now

**Review by atxr — APPROVED** [review]
